File: Tests/FunctionalTests/TestPyFit/testfit01.py
# functional test: two parameter fit using variety of minimizers
#
# In this test we are using simple geometry: cylinders without interference in
# air layer with two parameters (radius and height of cylinders), describing
# the sample. Our "real" data is 2D intensity map obtained from the simulation of
# the same geometry with fixed values height = 5nm and radius = 5nm.
# Then we run our minimization consequently using different minimization engines,
# with height=4nm, radius=6nm as starting fit parameter values.


import sys
import os
import numpy
import time

sys.path.append(os.path.abspath(
                os.path.join(os.path.split(__file__)[0],
                '..', '..', '..', 'lib')))

from libBornAgainCore import *
from libBornAgainFit import *

# sample parameters we are going to find
cylinder_height = 5*nanometer
cylinder_radius = 5*nanometer

# minimizer name and type of minimization algorithm
Minimizers = [ 
    ("Minuit2","Migrad"), 
    ("Minuit2","Fumili"), 
    ("GSLMultiMin","BFGS"),
    ("GSLMultiMin","SteepestDescent"),
    ("GSLMultiFit",""),
#    ("GSLSimAn","")
]


# -----------------------------------------------------------------------------
# run several minimization rounds using different minimizers
# -----------------------------------------------------------------------------
def runTest():
  #print "**********************************************************************"
  #print "*  Starting  TestFit01                                               *"
  #print "**********************************************************************"
  nTest=0
  status = "OK"
  for m in Minimizers:
    minimizer_name = m[0]
    minimizer_algorithm = m[1]
    print "Minimizer {0:-2d}   {1:}({2:})".format(nTest, minimizer_name, minimizer_algorithm)
    result_ok = run_fitting(minimizer_name, minimizer_algorithm)
    nTest+=1
    if not result_ok: status = "FAILED"

  return "TestFit01", "Two parameters fit using variety of minimizers.", status


# -----------------------------------------------------------------------------
# run fitting specified minimizer
# -----------------------------------------------------------------------------
def run_fitting(minimizer_name, minimizer_algorithm):
  sample = buildSample()
  simulation = createSimulation()
  simulation.setSample(sample)

  # creating real data, which is simply results of our simulation with default values
  simulation.runSimulation()
  real_data = simulation.getOutputDataClone()

  # setting fit suite
  fitSuite = FitSuite()
  fitSuite.setMinimizer( MinimizerFactory.createMinimizer(minimizer_name, minimizer_algorithm) )
  fitSuite.addFitParameter("*height", 4.*nanometer, 0.04*nanometer, AttLimits.lowerLimited(0.01) )
  fitSuite.addFitParameter("*radius", 6.*nanometer, 0.06*nanometer, AttLimits.lowerLimited(0.01) )
  fitSuite.addSimulationAndRealData(simulation, real_data)

  # run fit
  start_time = time.time()
  fitSuite.runFit()
  real_time = time.time() - start_time

  height_found = fitSuite.getMinimizer().getValueOfVariableAtMinimum(0)
  height_diff = abs(height_found - cylinder_height)/cylinder_height
  radius_found = fitSuite.getMinimizer().getValueOfVariableAtMinimum(1)
  radius_diff = abs(radius_found - cylinder_radius)/cylinder_radius

  print "            RealTime : {0:.3f} sec".format(real_time)
  print "            NCalls   : {0:<5d}".format(fitSuite.getNCalls())
  print '            par1     : {0:.4f} ({1:.3g}) '.format(height_found, height_diff)
  print '            par2     : {0:.4f} ({1:.3g}) '.format(radius_found, radius_diff)

  diff = 1.0e-02
  isSuccess = True
  if( (height_diff > diff) or (radius_diff > diff) ) : isSuccess=False
  return isSuccess



# -----------------------------------------------------------------------------
# create cylinders in the air
# -----------------------------------------------------------------------------
def buildSample():
    cylinder_ff = FormFactorCylinder(cylinder_height, cylinder_radius)
    n_particle = complex(1.0-6e-4, 2e-8)
    cylinder = Particle(n_particle, cylinder_ff)
    interference = InterferenceFunctionNone()

    particle_decoration = ParticleDecoration()
    particle_decoration.addParticle(cylinder)
    particle_decoration.addInterferenceFunction(interference)

    mAmbience = MaterialManager.getHomogeneousMaterial("Air", 1.0, 0.0 )
    air_layer = Layer(mAmbience)
    air_layer_decorator = LayerDecorator(air_layer, particle_decoration)
    multi_layer = MultiLayer()
    multi_layer.addLayer(air_layer_decorator)

    return multi_layer


def createSimulation():
    simulation = Simulation();
    simulation.setDetectorParameters(100, 0.0*degree, 2.0*degree,100 , 0.0*degree, 2.0*degree);
    simulation.setBeamParameters(1.0*angstrom, -0.2*degree, 0.0*degree);
    simulation.setBeamIntensity(1e10);
    return simulation


#-------------------------------------------------------------
# main()
#-------------------------------------------------------------
if __name__ == '__main__':
  name,description,status = runTest()
  print name,description,status

File: Tests/FunctionalTests/TestFit/TestFit01/TestFit01.cpp
#include "TestFit01.h"
#include "MaterialManager.h"
#include "MessageService.h"
#include "MultiLayer.h"
#include "LayerDecorator.h"
#include "ParticleDecoration.h"
#include "InterferenceFunctions.h"
#include "FormFactorCylinder.h"
#include "Units.h"
#include "FitSuite.h"
#include "MinimizerFactory.h"

#include "TBenchmark.h"

#include <iostream>
#include <iomanip>
#include <boost/format.hpp>

using namespace FunctionalTests;


TestFit01::TestFit01()
    : m_test_name("TestFit01")
    , m_test_description("Two parameters fit using variety of minimizers.")
    , m_cylinder_height(5.0*Units::nanometer)
    , m_cylinder_radius(5.0*Units::nanometer)
{
    m_minimizers.push_back( Minimizer("Minuit2","Migrad") );
    m_minimizers.push_back( Minimizer("Minuit2","Fumili") );
    m_minimizers.push_back( Minimizer("GSLMultiMin","BFGS") );
    m_minimizers.push_back( Minimizer("GSLMultiMin","SteepestDescent") );
    m_minimizers.push_back( Minimizer("GSLMultiFit","") ); // this is Levenberg-Marquard
//    m_minimizers.push_back( Minimizer("GSLSimAn","") );
//    m_minimizers.push_back( Minimizer("Genetic","") );
}


TestFit01::~TestFit01()
{

}

// run several minimization rounds using different minimizers
int TestFit01::run()
{
    std::cout << "*************************************"
            "*********************************" <<std::endl;
    std::cout << "*  Starting  TestFit01               "
            "                                *" <<std::endl;
    std::cout << "*************************************"
            "*********************************" <<std::endl;

    bool status_ok(true);
    for(size_t i=0; i<m_minimizers.size(); ++i ) {
        std::cout << boost::format("Test %1%: %|5t| %2%(%3%)\n") % i %
                m_minimizers[i].name % m_minimizers[i].algorithm;

        bool result = run_fitting(m_minimizers[i].name, m_minimizers[i].algorithm);
        status_ok &= result;
    }

    std::cout << m_test_name << " " << m_test_description << " " <<
            (status_ok ? "[OK]" : "[FAILED]") << std::endl;
    return (status_ok ? 0 : 1);
}


// run fitting using dedicated minimizer, return true in the case of success
bool TestFit01::run_fitting(const std::string &minimizer_name,
        const std::string &minimizer_algorithm)
{
    TBenchmark mb;
    mb.Start("test");

    ISample *sample = buildSample();
    Simulation *simulation = createSimulation();
    simulation->setSample(*sample);

    // Creating real data, which is simply results of our simulation with default values
    simulation->runSimulation();
    OutputData<double> *real_data = simulation->getOutputDataClone();

    // setting fitting
    FitSuite *fitSuite = new FitSuite();
    fitSuite->setMinimizer( MinimizerFactory::createMinimizer(
            minimizer_name, minimizer_algorithm) );
    fitSuite->addFitParameter("*height", 4.*Units::nanometer,
            0.04*Units::nanometer, AttLimits::lowerLimited(0.01) );
    fitSuite->addFitParameter("*radius", 6.*Units::nanometer,
            0.06*Units::nanometer, AttLimits::lowerLimited(0.01) );
    fitSuite->addSimulationAndRealData(*simulation, *real_data);

    // run fit
    fitSuite->runFit();

    double height_found = fitSuite->getMinimizer()->getValueOfVariableAtMinimum(0);
    double height_diff = std::fabs(height_found - m_cylinder_height)/m_cylinder_height;
    double radius_found = fitSuite->getMinimizer()->getValueOfVariableAtMinimum(1);
    double radius_diff = std::fabs(radius_found - m_cylinder_radius)/m_cylinder_radius;

    mb.Stop("test");
    std::cout << boost::format("%|12t| %-10s : %-6.3f \n") % "RealTime" %
            mb.GetRealTime("test");
    std::cout << boost::format("%|12t| %-10s : %-6.3f \n") % "CpuTime" %
            mb.GetCpuTime("test");
    std::cout << boost::format("%|12t| %-10s : %-4d \n") % "NCalls" %
            fitSuite->getNCalls();
    std::cout << boost::format("%|12t| %-10s : %-6.4f (diff %6.4g) \n") %
            "Par1" % height_found % height_diff;
    std::cout << boost::format("%|12t| %-10s : %-6.4f (diff %6.4g) \n") %
            "Par2" % radius_found % radius_diff;

    delete sample;
    delete simulation;
    delete real_data;
    delete fitSuite;

    const double diff = 1.0e-02;
    bool isSuccess = true;
    if( (height_diff > diff) || (radius_diff > diff) ) isSuccess=false;
    return isSuccess;
}


// create cylinders in the air
ISample *TestFit01::buildSample()
{
    MultiLayer *multi_layer = new MultiLayer();
    const IMaterial *air_material =
            MaterialManager::getHomogeneousMaterial("Air", 0.0, 0.0);
    Layer air_layer(air_material);
    complex_t n_particle(1.0-6e-4, 2e-8);
    ParticleDecoration particle_decoration( new Particle(n_particle,
            new FormFactorCylinder(m_cylinder_height, m_cylinder_radius)));
    particle_decoration.addInterferenceFunction(new InterferenceFunctionNone());
    LayerDecorator air_layer_decorator(air_layer, particle_decoration);
    multi_layer->addLayer(air_layer_decorator);
    return multi_layer;
}


// create simulation
Simulation *TestFit01::createSimulation()
{
    Simulation *simulation = new Simulation();
    simulation->setDetectorParameters(100, 0.0*Units::degree, 2.0*Units::degree,
            100 , 0.0*Units::degree, 2.0*Units::degree);
    simulation->setBeamParameters(1.0*Units::angstrom, 0.2*Units::degree,
            0.0*Units::degree);
    simulation->setBeamIntensity(1e10);
    return simulation;
}


#ifdef STANDALONE
int main()
{
//    MSG::SetLevel(MSG::INFO);
    FunctionalTests::TestFit01 test;
    return test.run();
}
#endif

